### PR TITLE
feat: wallet and treasury stats

### PR DESCRIPTION
**Pull request by BobTheBuidler**

### CODE REVIEW

**Review by banteg — CHANGES_REQUESTED** [review]

File: yearn/events.py [hunk 1]
```diff
@@ -23,6 +23,7 @@ def decode_logs(logs) -> EventDict:
         setattr(decoded[i], "block_number", log["blockNumber"])
         setattr(decoded[i], "transaction_hash", log["transactionHash"])
         setattr(decoded[i], "log_index", log["logIndex"])
+        setattr(decoded[i], "address", log["address"])
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/events.py]: done

File: yearn/treasury/treasury.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+
+def get_token_from_event(event):
+    try:
+        return event['Transfer'][0].address
+    except EventLookupError:
+        logger.critical(
+            f'One of your cached contracts has an incorrect definition: {event.address}. Please fix this manually'
+        )
+        raise (
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/treasury/treasury.py]: fixed

File: yearn/prices/constants.py [hunk 3]
```diff
@@ -18,4 +18,12 @@
     "0x1c48f86ae57291F7686349F12601910BD8D470bb": "usdk",
     "0x056Fd409E1d7A124BD7017459dFEa2F387b6d5Cd": "gusd",
     "0x0E2EC54fC0B509F445631Bf4b91AB8168230C752": "linkusd",
+    "0xa47c8bf37f92aBed4A126BDA807A7b7498661acD": "ust",
+    "0xe2f2a5C287993345a840Db3B0845fbC70f5935a5": "musd",
+    "0x5BC25f649fc4e26069dDF4cF4010F9f706c23831": "dusd",
+    "0xdF574c24545E5FfEcb9a659c229253D4111d87e1": "husd",
+    "0x99D8a9C45b2ecA8864373A26D1459e3Dff1e17F3": "mim",
+    "0x196f4727526eA7FB1e17b2071B3d8eAA38486988": "rsv",
+    "0x03ab458634910AaD20eF5f1C8ee96F1D6ac54919": "rai",
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/prices/constants.py]: fixed

File: yearn/__init__.py [hunk 4]
```diff
@@ -2,4 +2,8 @@
 from yearn.middleware import setup_middleware
 
 setup_logging()
-setup_middleware()
+try:
+    setup_middleware()
+except:
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/__init__.py]: fixed root issue and reverted

File: yearn/prices/magic.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ class PriceError(Exception):
 
 
 @ttl_cache(10000)
-def get_price(token, block=None):
+def get_price(token, block=None, silent=False):
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/prices/magic.py]: done

File: yearn/events.py [hunk 6]
```diff
@@ -31,6 +32,11 @@ def create_filter(address, topics=None):
     Create a log filter for one or more contracts.
     Set fromBlock as the earliest creation block.
     """
+    if address is None:
+        # We use this to pull Treasury's ERC20 Transfers
+        # Treasury didn't exist prior to block 10502337
+        return web3.eth.filter({"fromBlock": 10502337, "topics": topics})
```
[the remark this answers is not in the document]
Comment (BobTheBuidler) [reply yearn/events.py]: done

File: yearn/treasury/treasury.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    Used to export Yearn financial reports
+    '''
+
+    def __init__(self, watch_events_forever=False):
+        self.addresses = list(TREASURY_WALLETS)
+        self._transfers = []
+        self._topics_in = [
+            '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
+            None,
+            ['0x000000000000000000000000' + address[2:] for address in self.addresses],
```
Comment (banteg) [inline yearn/treasury/treasury.py:85]: alternatively here you can use `eth_utils.encode_hex(eth_abi.encode_single('address', address)`

File: yearn/treasury/treasury.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+        proxy_registry = Contract('0x4678f0a6958e4D2Bc4F1BAF7Bc52E8F3564f3fE4')
+        cdp_manager = Contract('0x5ef30b9986345249bc32d8928B7ee64DE9435E39')
+        # ychad = Contract('ychad.eth')
+        ychad = Contract('0xfeb4acf3df3cdea7399794d0869ef76a6efaff52')
+        vat = Contract('0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B')
+        proxy = proxy_registry.proxies(ychad)
+        cdp = cdp_manager.first(proxy)
+        urn = cdp_manager.urns(cdp)
+        ilk = encode_single('bytes32', b'YFI-A')
+        ink = vat.urns(ilk, urn, block_identifier=block).dict()["ink"]
```
Comment (banteg) [inline yearn/treasury/treasury.py:179]: btw you can access struct members even without calling `.dict()`

File: yearn/treasury/treasury.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+    def process_transfers(self, logs):
+        for log in logs:
+            try:
+                event = decode_logs(
+                    [log]
+                )  # NOTE: We have to decode logs here because NFTs prevent us from batch decoding logs
+                self._transfers.append(event)
+            except:
+                if log.address in [
+                    # These are NFTs # TODO
```
Comment (banteg) [inline yearn/treasury/treasury.py:298]: maybe there could be some heuristic to detect nfts?

**Review by banteg — APPROVED** [review]

File: yearn/treasury/treasury.py (hunk 9 above)
Comment (BobTheBuidler) [reply yearn/treasury/treasury.py:298]: Agreed, I had wanted to see a few more instances of NFTs that cause problems before I try to extrapolate from the way these particular logs are formed

**Pull request merged**